File: algorithms/2_medium/C/02257_count_unguarded_cells_in_the_grid/solution_2.cpp
class Solution {
public:
    int countUnguarded(int m, int n, vector<vector<int>>& guards, vector<vector<int>>& walls) {
        const size_t rows{static_cast<size_t>(m)};
        const size_t cols{static_cast<size_t>(n)};

        vector<vector<char>> grid(rows, vector<char>(cols, '.'));

        for (const auto &guard: guards) {
            grid[static_cast<size_t>(guard[0])][static_cast<size_t>(guard[1])] = 'g';
        }

        for (const auto &wall: walls) {
            grid[static_cast<size_t>(wall[0])][static_cast<size_t>(wall[1])] = 'w';
        }

        for (const auto& guard : guards) {
            const size_t x{static_cast<size_t>(guard[0])};
            const size_t y{static_cast<size_t>(guard[1])};

            for (size_t i {x - 1}; i < rows && grid[i][y] !='g' && grid[i][y] != 'w'; --i) {
                grid[i][y] = 'c';
            }

            for (size_t i {x + 1}; i < rows && grid[i][y] !='g' && grid[i][y] != 'w'; ++i) {
                grid[i][y] = 'c';
            }

            for (size_t j{y - 1}; j < cols && grid[x][j] != 'g' && grid[x][j] != 'w'; --j) {
                grid[x][j] = 'c';
            }

            for (size_t j{y + 1}; j < cols && grid[x][j] != 'g' && grid[x][j] != 'w'; ++j) {
                grid[x][j] = 'c';
            }
        }

        int count{0};
        for (size_t i{0}; i < rows; ++i) {
            for (size_t j{0}; j < cols; ++j) {
                if (grid[i][j] == '.') {
                    count += 1;
                }
            }
        }

        return count;
    }
};

File: algorithms/2_medium/C/02257_count_unguarded_cells_in_the_grid/solution_2.rs
impl Solution {
    pub fn count_unguarded(m: i32, n: i32, guards: Vec<Vec<i32>>, walls: Vec<Vec<i32>>) -> i32 {
        let rows = m as usize;
        let cols = n as usize;

        let mut grid = vec![vec![b'.'; cols]; rows];

        for guard in guards.iter() {
            grid[guard[0] as usize][guard[1] as usize] = b'g';
        }

        for wall in walls.into_iter() {
            grid[wall[0] as usize][wall[1] as usize] = b'w';
        }

        for guard in guards.into_iter() {
            let x = guard[0] as usize;
            let y = guard[1] as usize;

            for i in (0..x).rev() {
                if grid[i][y] == b'g' || grid[i][y] == b'w' {
                    break;
                }

                grid[i][y] = b'c';
            }

            for i in (x + 1)..rows {
                if grid[i][y] == b'g' || grid[i][y] == b'w' {
                    break;
                }

                grid[i][y] = b'c';
            }

            for j in (0..y).rev() {
                if grid[x][j] == b'g' || grid[x][j] == b'w' {
                    break;
                }

                grid[x][j] = b'c';
            }

            for j in (y + 1)..cols {
                if grid[x][j] == b'g' || grid[x][j] == b'w' {
                    break;
                }

                grid[x][j] = b'c';
            }
        }

        let mut count = 0_i32;
        for i in 0..rows {
            for j in 0..cols {
                if grid[i][j] == b'.' {
                    count += 1;
                }
            }
        }

        count
    }
}
